### PR TITLE
Rewrite for differential fetch plus other improvements

### PR DESCRIPTION
**Pull request by timabell**
To eliminate overlap with traditional paging which is to be removed.
Todo: Need to update individual call documentation.

### Context

Traditional pagination is going away after discussing with @misaka and being replaced with date+id based paging & incremental fetch

### Changes proposed in this pull request

Update first section of docs with new plan.

### Guidance to review

View rendered change here https://github.com/DFE-Digital/manage-courses-backend/blob/no-paging-doc/docs/api.md

Turn off whitespace when viewing diff. Maybe use kdiff to see full diff.

### CODE REVIEW

**Comment by timabell** [comment]
[Implement incremental fetch for providers.](https://trello.com/c/HMAlga4l/852-implement-incremental-fetch-for-providers)

File: docs/api.md [hunk 1]
```diff
@@ -21,125 +21,64 @@ curl "api_endpoint_here"
...
 
-## Changed-Records Pagination
+1. Call the endpoint with no query parameters,
+   e.g. `GET https://.../api/v1/<recruitment_cycle>/providers`
+2. The API will return the first page of records, and will include a response
+   header indicating the url needed to request the next page of records.
+3. Make another GET using the provided next page url.
+4. Repeat until no data is returned. Keep a copy of the next-page url
+   provided along with the empty response in order to be able to fetch
+   records that chanch after this initial load.
```
Comment (misaka) [inline docs/api.md]: "chanch" -> "change"

File: docs/api.md [hunk 2]
```diff
@@ -21,125 +21,64 @@ curl "api_endpoint_here"
...
-https://manage-courses-backend.herokuapp.com/api/v1/2018/courses?changed_since=20190114T223403Z
-```
+## Retrieving changed records
 
-When changed records are being retrieved, pagination is done by re-using the
-`changed_since` parameter. This pagination is state-less and any records that
-change while the results are being retrieved will appear again in the results,
-so clients should be prepared to receive duplicate records.
+1. Make a GET request using the next-page url provided with the empty
+   response at the end of the last full or incrmental fetch.
```
Comment (misaka) [inline docs/api.md]: "incrmental" -> "incremental"

File: docs/api.md [hunk 3]
```diff
@@ -21,125 +21,64 @@ curl "api_endpoint_here"
...
+2. The API will return the first page of records, and will include a response
+   header indicating the url needed to request the next page of records.
+3. Make another GET using the provided next page url.
+4. Repeat until no data is returned. Keep a copy of the next-page url
+   provided along with the empty response in order to be able to fetch
+   records that chanch after this initial load.
```
Comment (misaka) [inline docs/api.md]: "chanch" -> "change"

File: docs/api.md [hunk 4]
```diff
@@ -21,125 +21,64 @@ curl "api_endpoint_here"
...
-The response headers will include information related to continuing to retrieve
-changed records either immediately, or when ready to ingest more records.
-
-| Header       | Description                                                                                    |
-|--------------|------------------------------------------------------------------------------------------------|
-| Link         | Links to the next and last pages.                                                              |
-| Per-Page     | The maximum number of results on a page.                                                       |
-| Total        | The total number of results.                                                                   |
+**The query parameters are considered an interal concern of the API** and
+***must not** be constructed manually in order to avoid losing changes. The
```
Comment (misaka) [inline docs/api.md]: extra `*`

File: docs/api.md [hunk 5]
```diff
@@ -2,225 +2,207 @@
...
-|---------------|---------------------------------------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------|
-| changed_since | [ISO 8601 date/time string](https://en.wikipedia.org/wiki/ISO_8601) | Request all records that have changed since this date/time. Leave blank to retrieve all records                                       |
-| changed_id    | Integer                                                             | Optional. In the case where multiple records are changed within the same second, this is used to specify which record id to start at. |
-|               |                                                                     |                                                                                                                                       |
-
+**The query parameters are considered an interal concern of the API** and
```
Comment (defong) [inline docs/api.md]: interal

?

File: docs/api.md [hunk 6]
```diff
@@ -2,225 +2,207 @@
...
-As all responses from the API are paginated, a set maximum of results will be
-returned and a link to the next page will be included in the response headers.
-If the value for `Total` is less than `Per-Page` that indicates that there are
-no more results to retrieve with the `next` link, and if used 0 records will be
-returned. The `next` link may be saved for a future request to continue the
-incremental loading of changed data.
+* Bulk updates are happening to the data at the same time as a client is paging
+  through results.
+* More than a single page of records are updated within the same second.
+* Timestamps being generated for changes but being delayed in written to the
```
Comment (defong) [inline docs/api.md]: written

grammer?

File: docs/api.md [hunk 7]
```diff
@@ -2,225 +2,207 @@
...
+two sets of courses to manage – one set of courses that are currently
+published for the current recruitment cycle, and unpublished courses being
+preparated for the next recruitment cycle. Additionally, the providers who
+deliver next year's courses may change, and they may have different campuses
+for the same courses. The point in time when the overlap starts is referred
+to as rollover, and typically happens in or around May.
 
-To differentiate between entities from different recruitment cycles, each endpoint has a `<recruitment_cycle>` part in the URL. Additionally, the following entities have a `recruitment_cycle` attribute:
+To differentiate between entities from different recruitment cycles, each
+endpoint has a `<recruitment_cycle>` part in the URL. Additionally, the
```
Comment (defong) [inline docs/api.md]: does this still makes sense

File: docs/api.md [hunk 8]
```diff
@@ -394,54 +331,57 @@ curl "https://manage-courses-backend.herokuapp.com/api/v1/2019/subjects"
...
+region_code            | Text               | 01 to 11                                                                                                                                                                                                                                                                                                             | 2-character string
+data_download_format   | Text               | `ASCII DATA`, `ASCII DATA and UNICODE DATA` or `Unicode data`                                                                                                                                                                                                                                                        |
+oustanding_decisions   | Text               | `Alphabetic`, `Application code`, `Course / Alphabetic`, `Course / Application code`, `Course / Learner Number`, `Faculty / Alphabetic`, `Faculty / Application code`, `Faculty / Course / Alphabetic`, `Faculty / Course / Application code`, `Faculty / Course / Learner code`, `Not required` or `Learner Number` |
```
Comment (defong) [inline docs/api.md]: oustanding_decisions

?

Comment (timabell) [reply docs/api.md]: unchanged, I just reformatted the table, you could raise this separately

File: docs/api.md (hunk 7 above)
Comment (timabell) [reply docs/api.md]: adjusted slightly, but yes

**Pull request merged**